ingestion/functions/retrieval/retrieval.py

```diff
@@ -1,19 +1,66 @@
-import json
+import boto3
 import requests
+from datetime import datetime, timezone
+
+
+SOURCE_ID_FIELD = "sourceId"
+OUTPUT_BUCKET = "epid-raw-sources"
+TIME_FILEPART_FORMAT = "/%Y/%m/%d/%H%M/"
+
+s3_client = boto3.client("s3")
+
+
+def extract_source_id(event):
+    if SOURCE_ID_FIELD not in event:
+        error_message = (
+            "Required field {:s} not found in input event json.".format(
+                SOURCE_ID_FIELD))
+        print(error_message)
+        raise ValueError(error_message)
+    return event[SOURCE_ID_FIELD]
+
+
+def retrieve_content(source_id):
+    try:
+        ip = requests.get("http://checkip.amazonaws.com/").text.strip()
+        key_filename_part = "ip.txt"
+        file = open("/tmp/{:s}".format(key_filename_part), "w")
+        file.write(ip)
+        file.close()
+        s3_object_key = "{:s}{:s}{:s}".format(
+            source_id,
+            datetime.now(timezone.utc).strftime(TIME_FILEPART_FORMAT),
+            key_filename_part)
+        return (file.name, s3_object_key)
+    except requests.RequestException as e:
+        print(e)
+        raise e
+
+
+def upload_to_s3(file_name, s3_object_key):
+    try:
+        s3_client.upload_file(
+            file_name, OUTPUT_BUCKET, s3_object_key)
+        print("Uploaded source content to bucket {:s} with key {:s}".format(
```

Comment:
in python 3 I'd recommend fstrings to format instead of .format()

f"Uploaded source content to bucket {OUTPUT_BUCKET} with key {s3_object_key}"

that's similar to typescript f strings

Reply:
I'm living in the past! Updated, thanks.